api/src/handlers/list_validators.rs

```diff
@@ -172,16 +175,34 @@ pub async fn get_validators(
         },
     );
 
-    Ok(validators
-        .into_iter()
-        .skip(config.offset)
-        .take(config.limit)
-        .cloned()
-        .map(|v| ValidatorRecord {
-            epoch_stats: v.epoch_stats.into_iter().take(config.epochs).collect(),
-            ..v
-        })
-        .collect())
+    if let Some(from_data) = config.query_from_date {
```

Comment:
There is quite some overlap in the logic of the branches, please de-duplicate and refactor.

Reply:
Also, time to split this function to a multiple smaller ones, I think